Add api code (and scripts.py) to mypy type checking

pcustic opened the pull request:
## Description

Adding api code and `scripts.py` to `mypy` type checking in the Circulation Manager. Ran `mypy` on the excluded files and resolved issues.

## Motivation and Context
This is part 2 of the effort to have full type checking coverage in the CM:
https://www.notion.so/lyrasis/Get-Mypy-doing-type-checking-against-all-the-code-in-circulation-24f41995c1f74aec9963cb8b3d7b5238

## How Has This Been Tested?
Removed the files from the excluded list in the `pyproject.toml` and ran the `mypy`.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

jonathangreen commented:
@pcustic now that @io7m changes have gone in, there are some conflicts here. Can you resolve them, then we can get this code reviewed and in.

pcustic commented:
@jonathangreen conflicts resolved in https://github.com/ThePalaceProject/circulation/pull/652/commits/a05ac3daa283b4a1521c1670d7b4fdab574d97e5

pcustic commented on api/admin/controller/__init__.py:
```diff
@@ -1037,8 +1065,8 @@ def url_fn(after):
 
-    def custom_list(self, list_id: int) -> Union[Response, Dict, ProblemDetail]:
-        library = flask.request.library
+    def custom_list(self, list_id: int) -> Union[Response, Dict, ProblemDetail, None]:
```
Yeah, I would say it is, maybe a bit cleaner that way, I'll change it 👍

pcustic replied:
https://github.com/ThePalaceProject/circulation/pull/652/commits/eebf69c2c7b5e0c2ca6563ba4ee7a2e83b0f7284

pcustic commented on api/authenticator.py:
```diff
@@ -2632,14 +2632,6 @@ def remote_exchange_authorization_code_for_access_token(self, _db, code):
         """
         raise NotImplementedError()
 
-    def remote_patron_lookup(self, access_token):
```
On [line 2537](https://github.com/ThePalaceProject/circulation/pull/652/files/bc1393b235d9e2c1b7d8ab3e447672d8a581ea2a#diff-e4f8733d198e59b2bd6dc9532b19cb6387b9dfbe182f32474600bdbe9dc0ccbbR2537) there is method with same name and signature that is actually implemented :D though it only returns `None`

pcustic replied:
Yes, you are correct. I thought it should be the one that is implemented, but it should be in fact this one since that would be used if this method is called. Thanks 👍

pcustic commented on api/firstbook.py:
```diff
@@ -32,7 +33,7 @@ class FirstBookAuthenticationAPI(BasicAuthenticationProvider):
-    DEFAULT_IDENTIFIER_REGULAR_EXPRESSION = "^[A-Za-z0-9@]+$"
+    DEFAULT_IDENTIFIER_REGULAR_EXPRESSION = re.compile("^[A-Za-z0-9@]+$")
     DEFAULT_PASSWORD_REGULAR_EXPRESSION = "^[0-9]+$"
 
```
They're not worthy! :D

Just kidding, I'll check if I can add them to others. A bit lazy change on my part.

pcustic replied:
I removed `re.compile` since it is compiled in the `__init__` so there is no need for this and I changed so it's just strings everywhere.

pcustic commented on api/admin/controller/__init__.py:
```diff
@@ -1037,8 +1065,8 @@ def url_fn(after):
-    def custom_list(self, list_id: int) -> Union[Response, Dict, ProblemDetail]:
-        library = flask.request.library
+    def custom_list(self, list_id: int) -> Union[Response, Dict, ProblemDetail, None]:
+        library = flask.request.library  # type: ignore
```
Library isn't part of the flask, its something we inject in the `flask.request` in the [has_library decorator](https://github.com/ThePalaceProject/circulation/blob/main/api/routes.py#L157) or to be more precise, [here](https://github.com/ThePalaceProject/circulation/blob/main/api/base_controller.py#L155). For that reason flask types library doesn't know about it.

pcustic replied:
Added 👍

pcustic commented on api/s3_analytics_provider.py:
```diff
@@ -280,7 +280,7 @@ def _get_s3_uploader(self, db: Session) -> S3Uploader:
     @classmethod
-    def get_storage_settings(cls, db: Session) -> Dict:
+    def get_storage_settings(cls, db: Session) -> List[Dict]:
         """Return the provider's configuration settings including available storage options.
```
I mean, yeah, it returns list of dicts :D It seems to me it's the standard format for the CM settings. 

This uses `ConfigurationGrouping` `to_settings` method which returns settings in a list of dicts format - [here](https://github.com/ThePalaceProject/circulation/blob/main/core/model/configuration.py#L1399)

pcustic replied:
Yeah, we are able :) Thanks!

pcustic commented on api/sip/dialect.py:
```diff
@@ -6,7 +9,7 @@ class Dialect:
 
     # Settings defined in each class
-    sendEndSession = None
+    sendEndSession: Optional[bool] = None
```
Not sure, probably `None` means it's not set. It is used only [here](https://github.com/ThePalaceProject/circulation/blob/main/api/sip/client.py#L361).
Maybe have this explicitly check for `False`, not sure?

pcustic replied:
👍

pcustic commented on core/metadata_layer.py:
```diff
@@ -625,7 +625,7 @@ def __init__(
         checkouts_left: Optional[int] = None,
         terms_concurrency: Optional[int] = None,
-        content_types: Optional[str] = None,
+        content_types: Optional[List[str]] = None,
```
Hmm, I don't think there is any difference.

jonathangreen commented on core/model/collection.py:
```diff
@@ -867,7 +870,7 @@ def delete(self, search_index=None):
 
 
-collections_libraries = Table(
+collections_libraries: Table = Table(
```
Can you remove the `# type: ignore` on line 123, and add the same `: Table` annotation as you have here to `collections_identifiers` on line 890?

pcustic replied:
I knew there is one more instance for this situation but couldn't find it, thanks! Added!

jonathangreen reviewed (commented):
Looks good, just a few questions / comments.

pcustic commented on api/admin/controller/__init__.py:
```diff
@@ -801,8 +821,8 @@ def _list_as_json(self, list: CustomList, is_owner=True) -> Dict:
-    def custom_lists(self) -> Dict:
-        library = flask.request.library
+    def custom_lists(self) -> Union[Dict, ProblemDetail, Response, None]:
+        library = flask.request.library  # type: ignore  # "Request" has no attribute "library"
```
Added 👍

pcustic commented on api/admin/controller/library_settings.py:
```diff
@@ -5,7 +5,7 @@
 
 import flask
-import wcag_contrast_ratio
+import wcag_contrast_ratio  # type: ignore
```
No, not really :D Moved to `pyproject.toml`

pcustic commented on api/app.py:
```diff
@@ -55,7 +55,7 @@ def initialize_database(autoinitialize=True):
 
 from . import routes  # noqa
-from .admin import routes  # noqa
+from .admin import routes  # type: ignore  # noqa
```
Good idea, changed 👍

pcustic commented on api/circulation.py:
```diff
@@ -737,7 +739,7 @@ def __init__(
                 api = None
                 try:
-                    api = api_map[collection.protocol](db, collection)
+                    api = api_map[collection.protocol](db, collection)  # type: ignore
```
mypy understood that it was instance of `BaseCirculationAPI` (more precisely, it inherits from `BaseCirculationAPI`) however, there were no `__init__` for the`BaseCirculationAPI` with two arguments and mypy would show `Too many arguments for "BaseCirculationAPI"` error. I added the dummy `__init__(self, _db, collection)` that raises  `NotImplementedError()` for the `BaseCirculationAPI`. Let me know if you think that's OK?

pcustic commented on api/circulation.py:
```diff
@@ -749,10 +751,12 @@ def __init__(
-                ](collection)
+                ](
+                    collection
+                )  # type: ignore
```
Same as above just with `CirculationFulfillmentPostProcessor`

pcustic commented on api/clever/__init__.py:
```diff
@@ -105,7 +105,7 @@ class CleverAuthenticationAPI(OAuthAuthenticationProvider):
     # doesn't make sense for Clever. This removes the LIBRARY_SETTINGS from the
     # parent class.
-    LIBRARY_SETTINGS = []
+    LIBRARY_SETTINGS: list = []
```
Added 👍

pcustic commented on api/custom_index.py:
```diff
@@ -32,7 +34,7 @@ class CustomIndexView:
     """
 
-    BY_PROTOCOL = {}
+    BY_PROTOCOL: Dict[str, "CustomIndexView"] = {}
```
Thanks! 👍

pcustic commented on api/custom_patron_catalog.py:
```diff
@@ -25,7 +27,7 @@ class CustomPatronCatalog:
     """
 
-    BY_PROTOCOL = {}
+    BY_PROTOCOL: Dict[str, "CustomPatronCatalog"] = {}
```
Thanks 👍

pcustic commented:
@jonathangreen thanks for the comments added changes in https://github.com/ThePalaceProject/circulation/pull/652/commits/edf2b8ecd8daa4ae3804e6edb3888744226c3fef.

pcustic commented:
@jonathangreen There were some failed tests with latest commits so I fixed those. Regarding that change with adding dummy `__init__` of `BaseCirculationApi` - all of the inherited classes have the constructor with two parameters but the base class is used directly in the tests so I added those dummy parameters to the tests. I'm not sure that is the best solution since we are adding dummy code to the class and the tests just to pass the tests and type checking. Let me know what you think.

io7m commented:
> I'm not sure that is the best solution since we are adding dummy code to the class and the tests just to pass the tests and type checking. Let me know what you think.

I'm personally fine with this. I see this as a good example of the application of types exposing an inconsistency in the codebase.

io7m commented on api/authenticator.py:
```diff
@@ -2817,6 +2808,6 @@ class BaseSAMLAuthenticationProvider(
+    SETTINGS = SAMLSettings()  # type: ignore
 
-    LIBRARY_SETTINGS = []
+    LIBRARY_SETTINGS: List[dict] = []
```
Is the type of this `LIBRARY_SETTINGS` the same as the others below? There's one that has a stronger `List[Dict[str, Any]]` type.

pcustic replied:
Yeah, I think it is, [updated](https://github.com/ThePalaceProject/circulation/pull/652/commits/559aabf09aa167941018c7967c51d92aebb422ca)

pcustic commented:
@jonathangreen I updated this branch with the code from main, resolved all the conflicts and fixed errors that were introduced with the new code.

jonathangreen commented on api/circulation.py:
```diff
@@ -485,7 +485,12 @@ class BaseCirculationAPI:
+    ] = {}
+
+    def __init__(self, _db, collection):
+        pass
```
Should this raise 'NotImplementedError' as the instance below does?

pcustic replied:
Yeah, it makes sense. I think I first had both of them as `pass` but forgot to change this one. I'll change it 👍

pcustic replied:
> Should this raise 'NotImplementedError' as the instance below does?

As it turned out - no. Tests uses BaseCirculationAPI class directly so this raises an error and fails those tests. I'll revert back to pass.

jonathangreen reviewed (approved):
One minor question, but this looks good to go to me.
